util/network-devp2p/src/discovery.rs

```diff
@@ -591,65 +627,74 @@ impl<'a> Discovery<'a> {
+			self.expire_node_request(node_id);
 		}
-		removed
 	}
 
-	fn expire_in_flight_request(&mut self, node_id: NodeId, sent_at: Instant, removed: &mut HashSet<NodeId>) {
-		if let Entry::Occupied(entry) = self.in_flight_requests.entry(node_id) {
-			if entry.get().sent_at == sent_at {
-				entry.remove();
-
-				// Attempt to remove from bucket if in one.
-				let id_hash = keccak(&node_id);
-				let dist = Discovery::distance(&self.id_hash, &id_hash)
-					.expect("distance is None only if id hashes are equal; will never send request to self; qed");
-				let bucket = &mut self.node_buckets[dist];
-				if let Some(index) = bucket.nodes.iter().position(|n| n.id_hash == id_hash) {
-					if bucket.nodes[index].fail_count < self.request_backoff.len() {
-						let node = &mut bucket.nodes[index];
-						node.backoff_until = Instant::now() + self.request_backoff[node.fail_count];
-						node.fail_count += 1;
-						trace!(
-							target: "discovery",
-							"Requests to node {:?} timed out {} consecutive time(s)",
-							&node.address, node.fail_count
-						);
-					} else {
-						removed.insert(node_id);
-						let node = bucket.nodes.remove(index).expect("index was located in if condition");
-						debug!(target: "discovery", "Removed expired node {:?}", &node.address);
-					}
-				}
+	fn expire_node_request(&mut self, node_id: NodeId) {
+		// Attempt to remove from bucket if in one.
+		let id_hash = keccak(&node_id);
+		let dist = Discovery::distance(&self.id_hash, &id_hash)
+			.expect("distance is None only if id hashes are equal; will never send request to self; qed");
+		let bucket = &mut self.node_buckets[dist];
+		if let Some(index) = bucket.nodes.iter().position(|n| n.id_hash == id_hash) {
+			if bucket.nodes[index].fail_count < self.request_backoff.len() {
+				let node = &mut bucket.nodes[index];
+				node.backoff_until = Instant::now() + self.request_backoff[node.fail_count];
+				node.fail_count += 1;
+				trace!(
+					target: "discovery",
+					"Requests to node {:?} timed out {} consecutive time(s)",
+					&node.address, node.fail_count
+				);
+			} else {
+				let node = bucket.nodes.remove(index).expect("index was located in if condition");
+				debug!(target: "discovery", "Removed expired node {:?}", &node.address);
 			}
 		}
 	}
 
-	pub fn round(&mut self) -> Option<TableUpdates> {
-		let removed = self.check_expired(Instant::now());
-		self.discover();
-		if !removed.is_empty() {
-			Some(TableUpdates { added: HashMap::new(), removed })
-		} else { None }
+
+	pub fn round(&mut self) {
+		self.check_expired(Instant::now());
+		self.update_new_nodes();
+
+		if self.discovery_round.is_some() {
+			self.discover();
+		// Start discovering if the first pings have been sent (or timed out)
+		} else if self.in_flight_pings.len() == 0 && !self.discovery_initiated {
+			self.discovery_initiated = true;
+			self.refresh();
+		}
 	}
 
 	pub fn refresh(&mut self) {
-		self.start();
+		if self.discovery_round.is_none() {
```

Comment:
I don't really get it, whether it is fast or slow discovery timeout we end up on this same function : wouldn't this function be for fast only and slow discovery use the original one (with state clear and all).

Reply:
So _fast_ discovery is actually just calling this method more often, that's all.

Reply:
So both _fast_ and _normal_ call the same thing:
```rust
self.discovery.lock().as_mut().map(|d| d.refresh());
```

The `refresh` method won't actually do anything if there is already a discovery going on:
```rust
pub fn refresh(&mut self) {
	if self.discovery_round.is_none() {
		self.start();
	}
}
```

Does that answers your concerns?

Reply:
Not totally, but being rational and not looking at previous behaviour it seems better.

Reply:
Ah sorry, yeah the previous implementation was a bit borked at this level, but since it was called every minute, it was very unlikely that the previous discovery was still going on when called.